### ISSUE TITLE
Consider splitting 'Haplotype' into two subtypes

### ISSUE THREAD

**Issue by mbrush** [ga4gh-gks/variant-representation#15]
This is a proposal to consider splitting 'Haplotype' into two subtypes - **one to represent the unlocated** haplotype (which is a discontinuous set of alleles), and **the other to represent the located** version (which is a continuous extent of sequence that contains some variation as part).

This idea that a Haplotype can be these two very different things depending on whether a Location is provided seems to be a point of confusion for many. In my opinion explicitly separating these in different classes **makes this distinction easier to explain and understand**, and would **allow more precise specification supporting haplotype creation in the schema**.  
We already do polymorphism so it can be implemented with no technical barriers.

Another proposed benefit is that this **would make mapping to classes in variation-related ontologies possible** (e.g. SO, GENO MSO).  Each of these ontologies make high level distinctions between sequences that are continuous features and things that are discontinuous sets, and thus they hold no classes that could map to the current VMC haplotype definition.  But if haplotype is split as proposed, then mappings could be made.   For example, the unlocated vmc haplotype would map directly to a GENO:haplotype (which it defines as a set of discontinuous alterations in cis), and the located vmc haplotype would map to a GENO:complex allele . . . as GENO, like the ClinVar allele model, distinguishes 'simple' alleles that vary along their entire extent, from 'complex' alleles that are defined extent's of sequence that contain variation(s) as a proper part).

The ability to map to ontology terms **would let us point to an ontological-based, high-level, biologically-rooted, conceptual model of the VMC domain of discourse**.  And perhaps more importantly,  it would make it **easy for people to understand how vmc concepts are related to variation concepts in other models that are also mapped in GENO** - as GENO would act as a Roseetta Stone allowing this translation. This would be a big win w.r.t. **interoperability with models in other communities** (e.g. model organism/AGR) who are at this time also working to define variation models to support their data, and considering use/alignment with GENO concepts.

@reece @larrybabb @ahwagner curious as to your thoughts here.

**Comment by larrybabb** [comment]
This is the correct conceptual design and the rationale is well stated and crystal clear. I like the approach. 

I am still not super clear on how to weight the value proposition being made regarding the requirement to have a one to one mapping of classes to ontological concepts? 

Here’s why the decision is challenging when it seems like a no brainer...

Finding the proper balance between adding new concepts/classes to a messaging model to more accurately reflect the reality of the data versus the trade off of denormalizing parts of the schema in order to simplify the learning curve of adoption is the key.

I’m very confident that we would produce a less adoptable schema if it directly  represented the real world nature of the data being shared and we must first be diligent to not model aspects and concerns that are far beyond the needs of the the community. I feel like we do pretty good here. Second, I think there is utility in demoralizing (i.e.rolling up subclasses or embedding a dependent but optional class).

In this specific case I see the merit in simplifying a set of subclasses into a single concept with one optional location attribute.

If one wanted to store the shared data from this schema into an RDF based repo to perform reasoning then it seems reasonable to think that consumer could split out and associate the natural representation quite easily. But for those that simply want to share haplotyp data. They will not want to necessarily learn about why there are multiple types as it will be likely that they will be dealing with either one or the other type for a given service. I’m sure I can think of several other small but notable reasons for how and why implementors might be burdened by the additional subclass game for merely a single attribute (location) including initial perception.

I recognize that implementors will still need to understand and deal with the 2 different concepts or uses of the Haplotype schema but I do think the flatter the schema is the simpler it will be perceived and the more likely we will gain interest from engineers looking to investigate and try low threshold alternatives and solutions.

I do think this reveals an important point that has been hanging over us since day 1. I think it is vital that we maintain an information model that does deviate from these type of denormalizations. We need this for several reasons least of which may be keeping us from not loosing track of the core info model from which we denormalize. All the way to helping with providing documentation to show where the message schema originates from and how that maps to the correct semantic terms.

We need to add this tool to our approach for building and sustaining our ability to continue to extend and modify the schema for the foreseeable future. 

Allowing a place to demonstrate the proper semantic mappings to our schema for those tat need and want this is important and providing a simplified and reasonably denormalized schema will buy is much more users than otherwise.

**Comment by reece** [comment]
Totally agree with Larry's comments, which I would summarize this way: Splitting *might* be a fine idea, but we should keep it simple for now. Let's defer this question to the next version.

IMO, we're missing several prerequisites for a meaningful discussion:
* **Motivation.** Real-world use cases that demonstrate shortcomings of the current model to motivate the problem. By real-world, I mean a user story or similar that is missed or significantly impaired by the current Haplotype model. 
* **Proposal.** What should we do instead? What exactly is proposed?
* **Analysis.** We need an analysis of implementing diverse haplotype use cases using the current and proposed models. Our goal is to solve many uses cases well and simply, and that might make particular uses harder. We must make the system better overall, not just better for a specific concern.

Let's pick this topic up again after we've executed our current plan and when we have sufficient motivation to address it.

**Comment by mbrush** [comment]
Sounds like a plan guys, and thanks @larrybabb for the thoughtful analysis.  @rrfreimuth and I chatted the proposal at dinner the other night and will have some thoughts relevant to 'Analysis' when we revisit this question after the v1 release.

**Comment by ahwagner** [comment]
I agree with @mbrush re: the need for this change, and it is my opinion that our schema should reflect these two different concepts as separate types. In addition, I think we should construct formal representations of these concepts in our implementation to match the schema, but keep the _documentation_ low-barrier (with an emphasis on ease of adoption for common variation types).

Also, I would like to add my voice to the above consensus on tabling this discussion until we get the v1.0 schema out the door.
